### PR TITLE
[Runtime Env] asynchronous download and unzip

### PR DESCRIPTION
**Pull request by Catch-Bull**
Signed-off-by: Catch-Bull <burglarralgrub@gmail.com>

## Why are these changes needed?

Asynchronous download and unzip. The process of downloading and unzipping is run in a child process. The advantage of this is that `RuntimeEnvAgent` can stop these processes in time if it times out when it is created.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by architkulkarni — COMMENTED** [review]
Can you share more details about the timeout scenario you described?  When `setup_timeout_seconds` is reached, how does the download subprocess get stopped (does this part still need to be added, or does it already happen automatically in this PR?)

Should we also add a test where the runtime env download process gets stopped in the middle, and check that we can still successfully set up a runtime env after that?

File: python/ray/_private/runtime_env/packaging.py [hunk 1]
```diff
@@ -623,97 +626,54 @@ async def download_and_unpack_package(
...
-                unzip_package(
-                    package_path=pkg_file,
-                    target_dir=local_dir,
-                    remove_top_level_directory=True,
-                    unlink_zip=True,
-                    logger=logger,
+    local_dir = get_local_dir_from_uri(pkg_uri, base_directory)
+    assert local_dir != pkg_file, "Invalid pkg_file!"
+    if local_dir.exists():
+        assert local_dir.is_dir(), f"{local_dir} is not a directory"
```
Comment (SongGuyang) [inline python/ray/_private/runtime_env/packaging.py:637]:
```suggestion
        assert local_dir.is_dir(), f"Target directory {local_dir} already exists."
```

File: python/ray/_private/runtime_env/packaging.py [hunk 2]
```diff
@@ -623,97 +626,54 @@ async def download_and_unpack_package(
...
+    else:
+        protocol, pkg_name = parse_uri(pkg_uri)
+        if protocol == Protocol.GCS:
```
Comment (SongGuyang) [inline python/ray/_private/runtime_env/packaging.py:640]: Why we don't put the downloading logic of GCS to the downloader?

File: python/ray/_private/runtime_env/packaging.py [hunk 3]
```diff
@@ -623,97 +626,54 @@ async def download_and_unpack_package(
...
+            download_script = runtime_env_constans.DOWNLOAD_SCRIPT_FILENAME
+
+            download_cmd = [sys.executable, download_script]
+            download_cmd += ["--proto", protocol]
+            download_cmd += ["--src", pkg_uri]
+            download_cmd += ["--dest", pkg_file]
+
+            await check_output_cmd(download_cmd, logger=logger)
+
+            await unzip_package(
```
Comment (SongGuyang) [inline python/ray/_private/runtime_env/packaging.py:668]: If unzip failed, can we guarantee the downloaded file can be removed?

**Comment by architkulkarni** [comment]
Commenting to unstale

**Pull request closed without merging**